tests/helpers/test_condition.py

```diff
@@ -2837,6 +2837,128 @@ async def test_if_action_before_and_after_during(hass, hass_ws_client, calls):
     )
 
 
+async def test_if_action_before_or_after_during(hass, hass_ws_client, calls):
```

Comment:
To test this locally myself, I made a [parameterized test](https://docs.pytest.org/en/6.2.x/parametrize.html) so I could easily test out a bunch of variations with a single test, and I also just changed it to call `condition.sun` directly instead of creating the automation so the tests run faster.  

With a parameterized test, I suppose it doesn't matter whether the test is slightly longer with the automation components, but it could still help to simplify the test duplication.

```python
@pytest.mark.parametrize(
    "now_str, after, before, expected",
    [
        ( "2015-09-16 07:00:00+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Midnight
        ( "2015-09-16 12:33:18+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Sunrise - 1h
        ( "2015-09-16 13:32:18+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Sunrise - 1m
        ( "2015-09-16 13:33:17+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Sunrise - 1s
        ( "2015-09-16 13:33:18+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Sunrise
        ( "2015-09-16 13:33:19+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunrise + 1s
        ( "2015-09-16 13:34:18+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunrise + 1m
        ( "2015-09-16 14:33:18+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunrise + 1h
        ( "2015-09-17 00:53:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset - 1h
        ( "2015-09-17 01:52:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset - 1m
        ( "2015-09-17 01:53:44+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset - 1s
        ( "2015-09-17 01:53:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset
        ( "2015-09-17 01:53:46+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset + 1s
        ( "2015-09-17 01:54:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset + 1m
        ( "2015-09-17 02:53:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset + 1h
        ( "2015-09-17 06:59:59+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Midnight - 1s next day

        ( "2015-09-16 07:00:00+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Midnight
        ( "2015-09-16 12:33:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise - 1h
        ( "2015-09-16 13:32:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise - 1m
        ( "2015-09-16 13:33:17+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise - 1s
        ( "2015-09-16 13:33:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise
        ( "2015-09-16 13:33:19+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise + 1s
        ( "2015-09-16 13:34:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise + 1m
        ( "2015-09-16 14:33:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise + 1h
        ( "2015-09-17 00:53:45+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunset - 1h
        ( "2015-09-17 01:52:45+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunset - 1m
        ( "2015-09-17 01:53:44+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunset - 1s
        ( "2015-09-17 01:53:45+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Sunset
        ( "2015-09-17 01:53:46+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Sunset + 1s
        ( "2015-09-17 01:54:45+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Sunset + 1m
        ( "2015-09-17 02:53:45+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Sunset + 1h
        ( "2015-09-17 06:59:59+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Midnight - 1s next day

        ( "2015-09-16 07:00:00+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Midnight
        ( "2015-09-16 12:33:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunrise - 1h
        ( "2015-09-16 13:32:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunrise - 1m
        ( "2015-09-16 13:33:17+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunrise - 1s
        ( "2015-09-16 13:33:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunrise
        ( "2015-09-16 13:33:19+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunrise + 1s
        ( "2015-09-16 13:34:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunrise + 1m
        ( "2015-09-16 14:33:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunrise + 1h
        ( "2015-09-17 00:53:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunset - 1h
        ( "2015-09-17 01:52:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunset - 1m
        ( "2015-09-17 01:53:44+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunset - 1s
        ( "2015-09-17 01:53:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunset
        ( "2015-09-17 01:53:46+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunset + 1s
        ( "2015-09-17 01:54:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunset + 1m
        ( "2015-09-17 02:53:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunset + 1h
        ( "2015-09-17 06:59:59+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Midnight - 1s next day
    ],
)
async def test_condition_blahsun(hass, now_str, after, before, expected):
    """
    A test that verifies the behavior of the sun condition

    # sunrise: 2015-09-16 06:33:18 local, sunset: 2015-09-16 18:53:45 local
    # sunrise: 2015-09-16 13:33:18 UTC,   sunset: 2015-09-17 01:53:45 UTC
    """

    time_format = "%Y-%m-%d %H:%M:%S%z %Z"
    now = datetime.strptime(now_str, time_format)
    print(now.tzinfo)

    with patch("homeassistant.util.dt.utcnow", return_value=now):
        result = condition.sun(hass, before, after, None, None)
        assert result == expected

        # When we swap the before and after conditions, the result should be opposite
        result = condition.sun(hass, after, before, None, None)
        assert result != expected
```

Feel free to steal this if you think it's useful.  Could probably also get rid of a bunch of the extra cases (like -1h, -1m, +1m, +1h), I was just trying to be thorough but I don't think they're really necessary.

Reply:
@veleek  Did you test the "nighttime" condition as well? I don't see any call with "before sunrise, after sunset" but it may be simply an excerpt issue.

I'll gladly merge in your parameterized tests in my PR.

Reply:
That's actually covered by the "swap" portion of the test.  As I was duplicating test cases, I realized that it was essentially identical to just swapping before and after and swapping the expected result.  I couldn't think of a good way to name stuff or indicate that otherwise.  Could just duplicate the test cases if we wanted.

```python
    with patch("homeassistant.util.dt.utcnow", return_value=now):
        result = condition. Sun(hass, before, after, None, None)
        assert result == expected

        # When we swap the before and after conditions, the result should be opposite
        result = condition. Sun(hass, after, before, None, None)
        assert result != expected
```

Reply:
@veleek Can you submit a PR with your testcase changes?

Reply:
I'll try to do that in the next day or two if I have some time!